File: app/src/main/java/org/scp/gymlog/ui/common/dialogs/MenuDialogFragment.java
package org.scp.gymlog.ui.common.dialogs;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;

import androidx.annotation.MenuRes;
import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.scp.gymlog.R;

import java.util.function.Consumer;

public class MenuDialogFragment extends DialogFragment {
    public static final int DIALOG_CLOSED = -1;

    private final int menuId;
    private final Consumer<Integer> callback;
    private boolean callbackCalled = false;

    public MenuDialogFragment(@MenuRes int menuId, Consumer<Integer> callback) {
        this.menuId = menuId;
        this.callback = callback;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        LayoutInflater inflater = requireActivity().getLayoutInflater();
        View view = inflater.inflate(R.layout.dialog_menu, null);
        RecyclerView recyclerView = view.findViewById(R.id.parent_layout);

        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(new MenuRecyclerViewAdapter(getContext(), menuId,
                this::onMenuElementClicked));

        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity()).setView(view);

        return builder.create();
    }

    private void onMenuElementClicked(int menuItemId) {
        callback.accept(menuItemId);
        callbackCalled = true;
        dismiss();
    }

    @Override
    public void onDismiss(@NonNull DialogInterface dialog) {
        if (!callbackCalled) {
            callback.accept(DIALOG_CLOSED);
        }
        super.onDismiss(dialog);
    }
}

File: app/src/main/java/org/scp/gymlog/ui/common/dialogs/MenuDialogFragment.kt
package org.scp.gymlog.ui.common.dialogs

import android.app.AlertDialog
import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import androidx.annotation.MenuRes
import androidx.fragment.app.DialogFragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.scp.gymlog.R
import java.util.function.Consumer

class MenuDialogFragment(
    @param:MenuRes private val menuId: Int,
    private val callback: Consumer<Int>
) : DialogFragment() {

    companion object {
        const val DIALOG_CLOSED = -1
    }

    private var callbackCalled = false

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val inflater = requireActivity().layoutInflater
        val view = inflater.inflate(R.layout.dialog_menu, null)
        val recyclerView = view.findViewById<RecyclerView>(R.id.parentLayout)

        recyclerView.layoutManager = LinearLayoutManager(context)
        recyclerView.adapter = MenuRecyclerViewAdapter(requireContext(), menuId) { menuItemId ->
            onMenuElementClicked(menuItemId)
        }

        val builder = AlertDialog.Builder(activity).setView(view)
        return builder.create()
    }

    private fun onMenuElementClicked(menuItemId: Int) {
        callback.accept(menuItemId)
        callbackCalled = true
        dismiss()
    }

    override fun onDismiss(dialog: DialogInterface) {
        if (!callbackCalled) {
            callback.accept(DIALOG_CLOSED)
        }
        super.onDismiss(dialog)
    }
}
